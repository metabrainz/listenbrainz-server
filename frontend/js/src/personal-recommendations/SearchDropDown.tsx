import * as React from "react";
import { faUser } from "@fortawesome/free-solid-svg-icons";
import ListenControl from "../common/listens/ListenControl";

export type SearchDropDownProps = {
  action: (event: string) => void;
  suggestions?: Array<string>;
};

function SearchDropDown(props: SearchDropDownProps) {
  const { suggestions, action } = props;
  return (
    <div className="search-dropdown">
      {suggestions?.map((name) => {
        return (
          <ListenControl
            key={name}
            text={name}
            action={() => {
              action(name);
            }}
            icon={faUser}
          />
        );
      })}
    </div>
  );
}

export default SearchDropDown;
