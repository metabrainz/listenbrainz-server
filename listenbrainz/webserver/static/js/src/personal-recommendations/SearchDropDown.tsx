import * as React from "react";
import { faUser } from "@fortawesome/free-solid-svg-icons";
import ListenControl from "../listens/ListenControl";

export type SearchDropDownProps = {
  action: (event: string) => void;
  suggestions?: Array<string> | null;
};

const SearchDropDown = (props: SearchDropDownProps) => {
  const { suggestions, action } = props;
  return (
    <div className="searchdropdown">
      {suggestions!.map((name) => {
        return (
          <ListenControl
            text={name}
            action={() => {
              action(name);
            }}
            icon={faUser}
          />
        );
      })}
    </div>
  );
};

export default SearchDropDown;
