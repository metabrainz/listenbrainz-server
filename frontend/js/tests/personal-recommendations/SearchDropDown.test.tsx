import * as React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import SearchDropDown, {
  SearchDropDownProps,
} from "../../src/personal-recommendations/SearchDropDown";

const mockAction = jest.fn();
const props: SearchDropDownProps = {
  action: mockAction,
  suggestions: ["hrik2001", "riksucks"],
};

describe("SearchDropDown", () => {
  it("renders a list of suggestions as buttons", () => {
    render(<SearchDropDown {...props} />);

    const suggestionButtons = screen.getAllByRole("menuitem");
    expect(suggestionButtons).toHaveLength(2);
    expect(suggestionButtons.at(0)).toHaveTextContent("hrik2001");
    expect(suggestionButtons.at(1)).toHaveTextContent("riksucks");
  });

  it("calls the action with the correct name when a suggestion is clicked", async () => {
    const user = userEvent.setup();
    render(<SearchDropDown {...props} />);

    const firstSuggestion = screen.getByRole("menuitem", {
      name: /hrik2001/i,
    });
    await user.click(firstSuggestion);

    expect(mockAction).toHaveBeenCalledTimes(1);
    expect(mockAction).toHaveBeenCalledWith("hrik2001");
  });

  it("renders nothing if suggestions are not provided", () => {
    render(<SearchDropDown action={mockAction} suggestions={[]} />);
    expect(screen.queryByRole("button")).not.toBeInTheDocument();
  });
});
