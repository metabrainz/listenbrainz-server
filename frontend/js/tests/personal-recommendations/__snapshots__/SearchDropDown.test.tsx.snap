// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchDropDown /> renders correctly 1`] = `
<SearchDropDown
  action={[MockFunction]}
  suggestions={
    Array [
      "hrik2001",
      "riksucks",
    ]
  }
>
  <div
    className="searchdropdown"
  >
    <ListenControl
      action={[Function]}
      icon={
        Object {
          "icon": Array [
            448,
            512,
            Array [],
            "f007",
            "M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z",
          ],
          "iconName": "user",
          "prefix": "fas",
        }
      }
      text="hrik2001"
    >
      <button
        aria-label="hrik2001"
        disabled={false}
        onClick={[Function]}
        role="menuitem"
        title="hrik2001"
        type="button"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                448,
                512,
                Array [],
                "f007",
                "M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z",
              ],
              "iconName": "user",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-user fa-w-14 "
            data-icon="user"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </FontAwesomeIcon>
         
        hrik2001
      </button>
    </ListenControl>
    <ListenControl
      action={[Function]}
      icon={
        Object {
          "icon": Array [
            448,
            512,
            Array [],
            "f007",
            "M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z",
          ],
          "iconName": "user",
          "prefix": "fas",
        }
      }
      text="riksucks"
    >
      <button
        aria-label="riksucks"
        disabled={false}
        onClick={[Function]}
        role="menuitem"
        title="riksucks"
        type="button"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                448,
                512,
                Array [],
                "f007",
                "M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z",
              ],
              "iconName": "user",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-user fa-w-14 "
            data-icon="user"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </FontAwesomeIcon>
         
        riksucks
      </button>
    </ListenControl>
  </div>
</SearchDropDown>
`;
